Reset relocation/allocation failure counter on node join/shutdown

pxsalehi opened the pull request:
We prevent retries of allocations/relocations once they see `index.allocation.max_retries` failed attempts (default 5). In https://github.com/elastic/elasticsearch/pull/108987, we added reseting the allocation failure counters when a node joins the cluster. As discussed in the linked discussion, it would make sense to extend this reset also to relocations AND also consider node shutdown events. With this change we reset both allocation/relocation failures if a new node joins the cluster or a shutdown metadata is applied. The subset of shutdown events that we consider and how we track them is more or less copied from what was done for https://github.com/elastic/elasticsearch/pull/106998. To me the logic seemed to make sense here too. 

Closes ES-10492

pxsalehi committed:
reset both

pxsalehi commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -573,12 +578,43 @@ public void addAllocFailuresResetListenerTo(ClusterService clusterService) {
+     * Note that removing a non-RESTART shutdown metadata from a node that is still in the cluster is treated similarly and
+     * will cause resetting the allocation/relocation failures.
+     */
+    private boolean shouldResetAllocationFailures(ClusterChangedEvent changeEvent) {
```
In the PR allocation and relocation are kind of teated similarly and sometimes I've used "allocation" to mean both.

pxsalehi commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -573,12 +578,43 @@ public void addAllocFailuresResetListenerTo(ClusterService clusterService) {
     }
 
+    /**
+     *  We should reset allocation/relocation failure count to allow further retries when:
```
In the PR, we essentially reset when we know there might be shard movements caused by the events chosen here. For that reason, I'm not considering RESTART. Same logic as in https://github.com/elastic/elasticsearch/pull/106998.

pxsalehi replied:
I had an unsuccessful attempt at "generalizing" that logic to reuse here, but it ended up a bit weird. So I've just reused those few lines. When we reset the desired balance, and when we reset the allocation/relocation attempts, might not necessarily always share the same logic, I suppose, so not bad to have them separate IMO.

pxsalehi committed:
Update docs/changelog/119968.yaml

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

ywangd commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -573,12 +578,43 @@ public void addAllocFailuresResetListenerTo(ClusterService clusterService) {
+    private boolean shouldResetAllocationFailures(ClusterChangedEvent changeEvent) {
+        final var clusterState = changeEvent.state();
+        boolean hasAllocationFailures = clusterState.getRoutingNodes().hasAllocationFailures();
+        boolean hasRelocationFailures = clusterState.getRoutingNodes().hasRelocationFailures();
```
It might just be me not familiar with the code. But does the `reset-allocation-failures` task clear the relocation failures? Or does it not matter?

pxsalehi replied:
[Clears both](https://github.com/elastic/elasticsearch/blob/bedeb0414a7bdcfdca82cf78471c561bc979e7b0/server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java#L1301). As for the naming see [this comment](https://github.com/elastic/elasticsearch/pull/119968/files#r1916722015).

ywangd replied:
Ah ok thanks.

ywangd commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -573,12 +578,43 @@ public void addAllocFailuresResetListenerTo(ClusterService clusterService) {
+        boolean hasAllocationFailures = clusterState.getRoutingNodes().hasAllocationFailures();
+        boolean hasRelocationFailures = clusterState.getRoutingNodes().hasRelocationFailures();
+        var shutdownEventAffectsAllocation = false;
+        final var nodes = clusterState.nodes();
+        final var nodeShutdowns = clusterState.metadata().nodeShutdowns();
+        // If we remove a shutdown marker from a node, but it is still in the cluster, we could re-attempt failed relocations/allocations.
+        shutdownEventAffectsAllocation = processedNodeShutdowns.stream()
+            .anyMatch(nodeId -> nodeShutdowns.contains(nodeId) == false && nodes.get(nodeId) != null);
+        // Clean up processed shutdowns that are removed from the cluster metadata
+        processedNodeShutdowns.removeIf(nodeId -> nodeShutdowns.contains(nodeId) == false);
+        for (var shutdown : nodeShutdowns.getAll().entrySet()) {
+            // A RESTART doesn't necessarily move around shards, so no need to consider it for a reset.
+            // Furthermore, once the node rejoins after restarting, there will be a reset if necessary.
+            if (shutdown.getValue().getType() != SingleNodeShutdownMetadata.Type.RESTART) {
+                shutdownEventAffectsAllocation |= processedNodeShutdowns.add(shutdown.getKey());
+            }
+        }
+        return (changeEvent.nodesAdded() || shutdownEventAffectsAllocation) && (hasAllocationFailures || hasRelocationFailures);
```
Nit: can we short-circuit earlier when `nodesAdd()` is `true` which feels a bit easier to read to me?

pxsalehi replied:
We'd need to do the book-keeping of the seen shutdowns any way. I can add a comment or break the return line into multiple lines maybe. Frankly, I found this more readable. Maybe a comment-equivalent of that helps?

ywangd replied:
I wonder whether we can do away with the book-keeping? Can we compare the old and new cluster states and decide whether a shutdown record has been "added" or "removed but still in the cluster" and thus needing reset? It maybe helpful to enhance `ClusterChangedEvent` to provide some `shutdownChanged()` method similar to `nodesChanged()` so that it can be reused. This can be a separate PR. For the work here, a localized solution is fine.

pxsalehi replied:
Unless you feel strongly about this, I'd rather not revisit that part in this PR. We have used this elsewhere and it's edge cases has been reviewed (as mentioned in the PR description). So reusing it makes sense. I can follow up with your suggestion and if it works out also simplify the similar code we have for auto-resetting the desired balance. 

I can break off the return statement to something like 
```java
if (changeEvent.nodesAdded() || shutdownEventAffectsAllocation) {
  return hasAllocationFailures || hasRelocationFailures;
}
return false;
```

ywangd replied:
Yeah for consistency purpose, I am ok with this PR to keep the part as is. We can have a follow-up to then potentially change both places. 

To make sure we are not cross-talking, I am posting below the concrete code suggestion. To me the simplicity comes from no need for tracking shutdown nodes. I tested it with AllocationFailuresResetOnShutdownIT` and `AllocationFailuresResetIT` and they both both though I likely have not exhausted all possible test varaints. Let's put it through its own review process if you agree it's worthwhile to follow.

```java
    private boolean shouldResetAllocationFailures(ClusterChangedEvent changeEvent) {
        if (changeEvent.state().getRoutingNodes().hasAllocationFailures() == false
            && changeEvent.state().getRoutingNodes().hasRelocationFailures() == false) {
            return false;
        }
        if (changeEvent.nodesAdded()) {
            return true;
        }
        final var previous = nonRestartShutdownNodes(changeEvent.previousState());
        final var current = nonRestartShutdownNodes(changeEvent.state());
        if (previous.equals(current)) {
            return false;
        }
        return Sets.difference(current, previous).isEmpty() == false // new shutdown
            // removed shutdown but the node is still in cluster
            || Sets.difference(previous, current).stream().anyMatch(nodeId -> changeEvent.state().nodes().get(nodeId) != null);
    }

    // This can be a method on either ClusterState or NodesShutdownMetadata
    private static Set<String> nonRestartShutdownNodes(ClusterState clusterState) {
        return clusterState.metadata()
            .nodeShutdowns()
            .getAll()
            .values()
            .stream()
            .filter(m -> m.getType() != SingleNodeShutdownMetadata.Type.RESTART)
            .map(SingleNodeShutdownMetadata::getNodeId)
            .collect(Collectors.toUnmodifiableSet());
    }
```

ywangd commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -573,12 +578,43 @@ public void addAllocFailuresResetListenerTo(ClusterService clusterService) {
+        var shutdownEventAffectsAllocation = false;
+        final var nodes = clusterState.nodes();
+        final var nodeShutdowns = clusterState.metadata().nodeShutdowns();
+        // If we remove a shutdown marker from a node, but it is still in the cluster, we could re-attempt failed relocations/allocations.
+        shutdownEventAffectsAllocation = processedNodeShutdowns.stream()
+            .anyMatch(nodeId -> nodeShutdowns.contains(nodeId) == false && nodes.get(nodeId) != null);
```
This will cover restarted nodes as well. I think it's ok since it is covered by `nodesAdded` anyway. It may lead to reset a few more times:
1. On shutdown metadata (either restart or sigterm for serverless azure cluster)
2. On node join after restart
3. On shutdown metadata removal (if in a later cluster state update)

Still might be ok. Just to be explicit.

pxsalehi replied:
We explicitly avoid reset when the RESTART metadata is applied. When the node rejoins, yes, there will be a reset. We could explicitly prevent it, but not sure it is worth it. Removal of RESTART shouldn't cause a reset if the node is in the cluster. I think it is tested as well.

ywangd replied:
I should have been clear, by "restart", I meant the Azure specific behaviour where the node still goes through the sigterm shutdown record lifecyle while restarted by the Azure "auto-repair" feature. When it happens, we will reset the counter a few more times which should be ok.

ywangd reviewed (commented):
Haven't read the tests. Left a few minor comments. Also, I think we want to add some logs when the failure counters are 5 (max) and are resetted. IIUC from the conversation, this reaching max failures might be an indication of issues in some part of the system. We reset the counters to ease support burden but in the meantime it sorta hides the problems and logs should be helpful to make them discoverable.

pxsalehi commented:
> add some logs when the failure counters are 5

Yeah, we could. You mean the same way that we [log resetting desired balance ](https://github.com/elastic/elasticsearch/blob/bedeb0414a7bdcfdca82cf78471c561bc979e7b0/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java#L207)or including details of which shards had failures? The latter seems a bit verbose, but some summary of it maybe.

ywangd reviewed (commented):
> You mean the same way that we [log resetting desired balance ](https://github.com/elastic/elasticsearch/blob/bedeb0414a7bdcfdca82cf78471c561bc979e7b0/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java#L207)or including details of which shards had failures?

I was thinking more like the later otherwise it is hard to go back in the logs to find the actual shards. It would be great if the log message has a list of shards (potentially truncated if the list is too large) that have 5 failures. What do you think?

ywangd commented on x-pack/plugin/shutdown/src/internalClusterTest/java/org/elasticsearch/xpack/shutdown/AllocationFailuresResetOnShutdownIT.java:
```diff
@@ -0,0 +1,302 @@
+            assertBusy(() -> {
+                var stateAfterNodeJoin = internalCluster().clusterService().state();
+                var relocatedShard = stateAfterNodeJoin.routingTable().index("index1").shard(0).primaryShard();
+                assertThat(relocatedShard.relocationFailureInfo().failedRelocations(), Matchers.lessThan(maxAttempts));
```
Probably for my knowledge: Why does this assert `failedRelocations < maxAttemps` instead of `failedRelocations == 0`?

pxsalehi replied:
what we care about is the reset of the counter. so anything below max would do. I don't see a reason to be too strict here, since the test will be extra brittle if for whatever reason it takes more than one attempt to allocate/relocate.

ywangd commented on x-pack/plugin/shutdown/src/internalClusterTest/java/org/elasticsearch/xpack/shutdown/AllocationFailuresResetOnShutdownIT.java:
```diff
@@ -0,0 +1,302 @@
+            }));
+        // add shutdown to the new node
+        final var request = createRequest(randomFrom(SingleNodeShutdownMetadata.Type.values()), node2, randomIdentifier());
+        client().execute(PutShutdownNodeAction.INSTANCE, request).get();
```
Similar nit here. It applies to a few more places and I won't comment on all of them to avoid being too repetitive.
```suggestion
        safeGet(client().execute(PutShutdownNodeAction.INSTANCE, request));
```

ywangd reviewed (commented):
The PR looks good to me. I have only minor comments. I'll review it again if you are adding logging to it. Alternatively, I am also ready to approve it if you prefer loggings to be a separate PR. Please let me know. Thanks!

pxsalehi commented:
Thanks. Since we've already gotten into both points here, I'll just add it here and ping you. Didn't have time to get back to this.

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps250109-resetCounterOnShutdown

pxsalehi committed:
only rely on ClusterChangedEvent

pxsalehi committed:
add logging

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps250109-resetCounterOnShutdown

pxsalehi committed:
Use safeGet

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps250109-resetCounterOnShutdown

pxsalehi commented on server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java:
```diff
@@ -1298,14 +1312,47 @@ public boolean hasAllocationFailures() {
+        return false;
+    }
+
+    public void resetFailedCounter(RoutingAllocation allocation) {
```
I've chosen to add the logging here, to be able to produce some reliable information and also it integrates with the existing code of going through the failures. `SETTING_ALLOCATION_MAX_RETRY` is an index scoped setting so I'm not referring to any value, just picking those reaching max. Anything under max is not mentioned, as was the case before.

pxsalehi commented:
Simplified the prod code a bit in https://github.com/elastic/elasticsearch/pull/119968/commits/d4140e58ad983a83c55d6cad0403a2113496dafa and added logging in https://github.com/elastic/elasticsearch/pull/119968/commits/f6918debe6095d925888651dcfe7f1ee9ee547c3. The simplification doesn't apply to https://github.com/elastic/elasticsearch/pull/106998 since we don't have the `ClusterChangedEvent` like here.

ywangd commented on server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java:
```diff
@@ -1298,14 +1312,47 @@ public boolean hasAllocationFailures() {
+                    );
+                    if (failedAllocations >= maxRetry) {
+                        shardsWithMaxFailedAllocations++;
+                        if (topShardIdsWithFailedAllocations.size() <= MAX_SHARDS_IN_LOG_MSG) {
```
Nit: I think we can check the size along with `failedAllocations > 0` to short-circuit a bit earlier and avoid resolving the setting value.

pxsalehi replied:
I don't think so,`shardsWithMaxFailedAllocations` counts all of the ones that have reached max, while  `topShardIdsWithFailedAllocations` stores the shard IDs.

ywangd commented on server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java:
```diff
@@ -1326,6 +1373,20 @@ public void resetFailedCounter(RoutingChangesObserver routingChangesObserver) {
+                        );
+                        if (failedRelocations >= maxRetry) {
+                            shardsWithMaxFailedRelocations++;
+                            if (topShardIdsWithFailedRelocations.size() <= MAX_SHARDS_IN_LOG_MSG) {
+                                topShardIdsWithFailedRelocations.add(shardRouting.shardId());
+                            }
```
Similar nit: checking size before resolving setting value would be my preference.

ywangd commented on server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java:
```diff
@@ -60,6 +67,13 @@
 
+    private static final Logger logger = LogManager.getLogger(RoutingNodes.class);
+    public static final String RESET_FAILED_ALLOCATION_COUNTER_LOG_MSG =
+        "Resetting failure counter for %d shard(s) that have reached their max allocation retires (%s)";
+    public static final String RESET_FAILED_RELOCATION_COUNTER_LOG_MSG =
+        "Resetting failure counter for %d shard(s) that have reached their max relocation retries (%s)";
```
Nit:
```suggestion
        "Resetting failure counter for [%d] shard(s) that have reached their max allocation retires (%s)";
    public static final String RESET_FAILED_RELOCATION_COUNTER_LOG_MSG =
        "Resetting failure counter for [%d] shard(s) that have reached their max relocation retries (%s)";
```

ywangd reviewed (approved):
LGTM

Thanks for the iterations! Test cases look very solid 👍

pxsalehi commented on server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java:
```diff
@@ -1298,14 +1312,47 @@ public boolean hasAllocationFailures() {
+    }
+
+    public void resetFailedCounter(RoutingAllocation allocation) {
+        final var observer = allocation.changes();
```
I very much prefer the `observer` since it returns a `RoutingChangesObserver`.

pxsalehi commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into ps250109-resetCounterOnShutdown
